### ISSUE TITLE
InvalidArgument false positive: `self` param in trait-declared scope resolves to the custom builder class

### ISSUE THREAD

**Issue by alies-dev** [psalm/laravel-psalm-plugin#1031]
**Describe the bug**

When a model uses a custom Eloquent builder and a legacy scope is declared **in a trait** with a non-query parameter typehinted `self`, calling that scope on the builder instance emits a false-positive `InvalidArgument`. Psalm reports the expected type of the argument as the custom builder class instead of the model:

```
ERROR: InvalidArgument
Argument 1 of App\Builders\RankedPostBuilder::rankedabove expects App\Builders\RankedPostBuilder,
but App\Models\RankedPost provided (see https://psalm.dev/004)
```

The same scope declared directly on the model class (not via a trait) does not trigger the issue.

**Code to reproduce**

`app/Builders/RankedPostBuilder.php`:

```php
<?php

namespace App\Builders;

use App\Models\RankedPost;
use Illuminate\Database\Eloquent\Builder;

/** @extends Builder<RankedPost> */
final class RankedPostBuilder extends Builder
{
}
```

`app/Models/Concerns/ComparesRank.php`:

```php
<?php

namespace App\Models\Concerns;

use Illuminate\Database\Eloquent\Builder;

/** @psalm-require-extends \Illuminate\Database\Eloquent\Model */
trait ComparesRank
{
    /** @param Builder<self> $query */
    public function scopeRankedAbove(Builder $query, self $model): void
    {
        $query->whereKeyNot($model->getKey());
    }
}
```

`app/Models/RankedPost.php`:

```php
<?php

namespace App\Models;

use App\Builders\RankedPostBuilder;
use App\Models\Concerns\ComparesRank;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;

final class RankedPost extends Model
{
    use ComparesRank;

    /** @param Builder<self> $query (same signature, but declared directly on the model: no issue) */
    public function scopeRankedBelow(Builder $query, self $model): void
    {
        $query->whereKeyNot($model->getKey());
    }

    public function newEloquentBuilder($query): RankedPostBuilder
    {
        return new RankedPostBuilder($query);
    }
}
```

Analyzed snippet:

```php
<?php

use App\Models\RankedPost;

$post = new RankedPost();

RankedPost::query()->rankedAbove($post); // InvalidArgument: expects RankedPostBuilder, RankedPost provided
RankedPost::query()->rankedBelow($post); // OK (scope declared directly on the model)
```

**Analysis**

The chain for `RankedPost::query()->rankedAbove($post)`:

1. `ModelRegistrationHandler::registerHandlersForModel()` registers the `CustomBuilderMethodHandler` scope providers on the custom builder class (the #630 path).
2. `CustomBuilderMethodHandler::getScopeMethodParamsOnBuilder()` delegates to `ModelMethodHandler::getScopeParams()`, which returns `$codebase->methods->getMethodParams(...)` for `scopeRankedAbove` minus the leading `$query` param. The slicing itself is correct.
3. For trait methods, Psalm stores the `self` typehint unresolved and expands it at argument-check time against the class the called method belongs to. Since the params are served for `RankedPostBuilder::rankedabove`, `self` expands to `RankedPostBuilder` instead of the model, hence the false positive. Scopes declared directly on the model are unaffected because `self` is already resolved to the model class at scan time.

A `static` typehint, or `self`/`static` inside a docblock-only param type in a trait scope, should hit the same expansion problem.

Suggested fix: in `ModelMethodHandler::getScopeParams()`, expand each returned param type with `TypeExpander::expandUnion()` using the model class as the `self`/static context before handing the params to Psalm, so `self`, `static`, and generic types like `Builder<self>` in scope signatures always resolve against the model, regardless of which class the params provider is registered on.

**Impacted Versions**

```text
psalm/plugin-laravel  dev-master (e4a18f85)
vimeo/psalm           6.16.1
laravel/framework     13.14.0 (also reproduced on 12.x)
PHP                   8.4
```

**Psalm config**

```xml
<?xml version="1.0"?>
<psalm
    errorLevel="1"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns="https://getpsalm.org/schema/config"
    xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
>
    <projectFiles>
        <directory name="app" />
    </projectFiles>
    <plugins>
        <pluginClass class="Psalm\LaravelPlugin\Plugin" />
    </plugins>
</psalm>
```

**Additional context**

Single-file analysis (`vendor/bin/psalm app/Models/RankedPost.php`) may not reproduce the error; a full project run does. Worth keeping in mind for the regression test.
